### PR TITLE
removespecificItem and decrement quantity functionality added

### PR DESCRIPTION
**Pull request by harshvaghasiya-webelight**

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+    "cSpell.words": [
+        "birthdate",
+        "deletedata",
+        "updatedata",
+        "userid",
+        "viewuser"
+    ]
+}
```
Comment (anshs-webelight) [inline .vscode/settings.json]: no need to push this file to repo, add this file to .gitignore file

File: src/controller/cartController.js [hunk 2]
```diff
@@ -6,15 +6,15 @@ exports.addToCart = async (req, res) => {
...
     // const availableUser = await cartModel.exists({ userId });   returns only id;
 
-    // If User is avalable then Added Proudcuts to same cart other wise create new cart
+    // If User is available then Added Products to same cart other wise create new cart
     if (availableUser) {
       // To save the all userid which is saved in user's specific cart
-      const allProductIdAvilableInCart = availableUser.products.map(
-        (product) => product.productId.toString()
+      const allProductIdAvilableInCart = availableUser.products.map((product) =>
+        product.productId.toString()
```
Comment (anshs-webelight) [inline src/controller/cartController.js]: ` const allProductIdAvilableInCart = availableUser.products.map(({productId}) =>  productId.toString()`

File: src/routes/cartRouter.js [hunk 3]
```diff
@@ -10,6 +10,10 @@ route.post("/", cartController.addToCart);
 route.delete("/:userId", cartController.removeFromCart);
 
 // Remove the Specific Item From cart
+route.post("/removeSpecificItem", cartController.removeSpecificItem);
+
+// Decrement the quantity from cart
+route.post("decrementQuantity",cartController.decrementQuantity)
```
Comment (anshs-webelight) [inline src/routes/cartRouter.js]: use proper subroutes and why created `post` request for delete something endpoint

Comment (anshs-webelight) [reply src/routes/cartRouter.js]: `/removeSpecificItem` ==> `/items/{productId}

Comment (anshs-webelight) [inline src/routes/cartRouter.js]: PATCH / PUT : /cart/quantity/{productId}

**Review by anshs-webelight — CHANGES_REQUESTED** [review]

File: src/controller/cartController.js (hunk 2 above)
Comment (harshvaghasiya-webelight) [reply src/controller/cartController.js]: resolved

File: .vscode/settings.json (hunk 1 above)
Comment (harshvaghasiya-webelight) [reply .vscode/settings.json]: resolved

File: src/routes/cartRouter.js (hunk 3 above)
Comment (harshvaghasiya-webelight) [reply src/routes/cartRouter.js]: what i do hear in cart functionality we not using the midddleware and if we pass the product id from params we can not find the userid to find user specific cart

Comment (anshs-webelight) [reply src/routes/cartRouter.js]: just give you an example, you can take userId from body

Comment (harshvaghasiya-webelight) [reply src/routes/cartRouter.js]: resolved

Comment (harshvaghasiya-webelight) [reply src/routes/cartRouter.js]: ok

File: src/routes/cartRouter.js [hunk 4]
```diff
@@ -10,6 +10,10 @@ route.post("/", cartController.addToCart);
 route.delete("/:userId", cartController.removeFromCart);
 
 // Remove the Specific Item From cart
+route.patch("/items/:productId", cartController.removeSpecificItem);
+
+// Decrement the quantity from cart
+route.patch("/reduceQuantity/:productId",cartController.reduceQuantity)
```
Comment (anshs-webelight) [inline src/routes/cartRouter.js]: don't use camel case words for sub-routes, use "-" separated names like `reduce-quantity` etc.

**Review by anshs-webelight — CHANGES_REQUESTED** [review]

File: src/routes/cartRouter.js (hunk 4 above)
Comment (harshvaghasiya-webelight) [reply src/routes/cartRouter.js]: resolved

File: src/controller/cartController.js [hunk 5]
```diff
@@ -45,21 +46,136 @@ exports.addToCart = async (req, res) => {
...
 // View the user data from cart
 exports.findCart = async (req, res) => {
   try {
-    const cartData = await cartModel
-      .findOne({ userId: req.params.userId })
-      .lean();
+    const cartData = await cartModel.aggregate([
+      {
+        $match: {
+          userid: new mongoose.Types.ObjectId("65a76d78ca09b0052681193a"),
```
Comment (anshs-webelight) [inline src/controller/cartController.js]: why static id?

File: src/controller/cartController.js [hunk 6]
```diff
@@ -45,21 +46,136 @@ exports.addToCart = async (req, res) => {
...
+              product: {
+                _id: "$product._id",
+                name: "$product.name",
+                description: "$product.description",
+                price: "$product.price",
+              },
+            },
+          },
+        },
+      },
```
Comment (anshs-webelight) [inline src/controller/cartController.js]: no need to add, you can comment it out for your reference

File: src/controller/orderController.js [hunk 7]
```diff
@@ -9,25 +9,25 @@ exports.checkOut = async (req, res) => {
...
+    // Find the product details from the product Collection which id is in allproductId
+    const allProduct = await productModel.find({ _id: { $in: allproductId } });
 
     // Finding the Total Amount of All Product in cart
     let totalBill = 0;
     allProduct.forEach(({ _id, price }) => {
-      const matchingProduct = cartProducts.products.find(({ productId }) =>
-        _id.equals(productId)
+      const matchingProduct = cartProducts.products.find(({ productid }) =>
+        _id.equals(productid)
```
Comment (anshs-webelight) [inline src/controller/orderController.js]: why changed `productId` ==> `productid`?

File: src/controller/orderController.js [hunk 8]
```diff
@@ -37,14 +37,14 @@ exports.checkOut = async (req, res) => {
 
     // Creating record in order table for history
     await orderModel.create({
-      userId: req.params.userId,
+      userid: req.params.userid,
```
Comment (anshs-webelight) [inline src/controller/orderController.js]: why changed and used `userid` instead of `userId`

File: src/models/cart.js [hunk 9]
```diff
@@ -2,13 +2,13 @@ const mongoose = require("mongoose");
...
 const cartSchema = new schema({
-  userId: {
+  userid: {
     type: schema.Types.ObjectId,
     ref: "User",
   },
   products: [
     {
-      productId: {
+      productid: {
```
Comment (anshs-webelight) [inline src/models/cart.js]: why?

File: src/models/cart.js [hunk 10]
```diff
@@ -2,13 +2,13 @@ const mongoose = require("mongoose");
 const schema = mongoose.Schema;
 
 const cartSchema = new schema({
-  userId: {
+  userid: {
```
Comment (anshs-webelight) [inline src/models/cart.js]: revert it

File: src/models/cart.js [hunk 11]
```diff
@@ -20,4 +20,4 @@ const cartSchema = new schema({
   ],
 });
 
-module.exports = mongoose.model("Cart", cartSchema);
+module.exports = mongoose.model("cart", cartSchema);
```
Comment (anshs-webelight) [inline src/models/cart.js:23]: why changed?

File: src/routes/userRouter.js [hunk 12]
```diff
@@ -13,10 +13,10 @@ route.get("/logout", isAuthenticated, userController.logout);
...
 // Update User Data
-route.put("/:userId", isAuthenticated, userController.updatedata);
+route.put("/:userid", isAuthenticated, userController.updatedata);
 // Delete User Data
-route.delete("/:userId", isAuthenticated, userController.deletedata);
-// View Perticular user
-route.get("/:userId", isAuthenticated, userController.viewuser);
+route.delete("/:userid", isAuthenticated, userController.deletedata);
+// View Particular user
+route.get("/:userid", isAuthenticated, userController.viewuser);
```
Comment (anshs-webelight) [inline src/routes/userRouter.js]: why changed `userId`  to `userid` everyWhere?

**Review by anshs-webelight — CHANGES_REQUESTED** [review]

File: src/controller/cartController.js (hunk 5 above)
Comment (harshvaghasiya-webelight) [reply src/controller/cartController.js]: soory forget to update it it's just for testing purpose

File: src/controller/cartController.js (hunk 6 above)
Comment (harshvaghasiya-webelight) [reply src/controller/cartController.js]: resolved

File: src/controller/orderController.js (hunk 8 above)
Comment (harshvaghasiya-webelight) [reply src/controller/orderController.js]: resolved

File: src/models/cart.js (hunk 11 above)
Comment (harshvaghasiya-webelight) [reply src/models/cart.js:23]: you told model name should be on small case so

Comment (anshs-webelight) [reply src/models/cart.js:23]: I said to maintain consistency because 2 models have uppercase and 2 model has small letter that's should not happened

**Pull request merged**